### PR TITLE
Fetching the initial state of the sequencer

### PR DESCRIPTION
**Pull request by szynwelski**

### CODE REVIEW

File: src/utils/listener/block_downloader.go [hunk 1]
```diff
@@ -196,6 +196,7 @@ func (self *BlockDownloader) run() error {
 		}
 	}
 
+	self.Stop()
```
Comment (janekolszak) [inline src/utils/listener/block_downloader.go:199]: I don't think this is the right place to stop. If the intention is to stop the application after the last block is processed maybe it's better to create a task that calls Controller.Stop() after the last block is processed?

File: src/utils/tool/random.go [hunk 2]
```diff
@@ -3,15 +3,10 @@ package tool
 import (
 	crypto_rand "crypto/rand"
 	"math/rand"
-	"time"
 )
 
 const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 
-func init() {
-	rand.Seed(time.Now().UnixNano())
```
Comment (janekolszak) [inline src/utils/tool/random.go:12]: Why? Without this seed is always 1.

File: src/initsequencer/controller.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		WithMonitor(monitor).
+		WithInputChannel(nextFinishedBlock.Output).
+		WithBackoff(0, config.Syncer.TransactionMaxInterval).
+		WithInitStartHeight(db, model.SyncedComponentInteractions)
+
+	writer := NewWriter(config, sequencerRepoPath).
+		WithSequencerRepoPath(sequencerRepoPath).
+		WithDB(db).
+		WithInput(blockDownloader.Output)
+
```
Comment (janekolszak) [inline src/initsequencer/controller.go:47]: And here writer could emit struct{} on it's Output, and another task could close the Controller

File: src/initsequencer/next-finished-block.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+		First(&state, model.SyncedComponentInteractions).
+		Error
+	if err != nil {
+		return
+	}
+
+	self.Output <- &arweave.NetworkInfo{
+		Height: int64(state.FinishedBlockHeight) + 1,
+	}
+	close(self.Output)
```
Comment (janekolszak) [inline src/initsequencer/next-finished-block.go:46]: This is ok, but everywhere else we use WithOnStop(func(){close(self.Output)}) to close the channel.

File: src/cmd/init-sequencer.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+		case <-controller.CtxRunning.Done():
+		case <-applicationCtx.Done():
+		}
+
+		controller.StopWait()
+
+		return
+	},
+	PostRunE: func(cmd *cobra.Command, args []string) (err error) {
+		log := logger.NewSublogger("root-cmd")
```
[the remark this answers is not in the document]
Comment (szynwelski) [reply src/cmd/init-sequencer.go]: done

File: src/utils/listener/block_downloader.go (hunk 1 above)
Comment (szynwelski) [reply src/utils/listener/block_downloader.go]: done

File: src/utils/tool/random.go (hunk 2 above)
Comment (szynwelski) [reply src/utils/tool/random.go:12]: In Go starting from version `1.20`, the global random number generator is automatically seeded.
https://tip.golang.org/doc/go1.20

File: src/initsequencer/next-finished-block.go (hunk 4 above)
Comment (szynwelski) [reply src/initsequencer/next-finished-block.go]: done

File: src/initsequencer/writer.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	input             chan *arweave.Block
+	Output            chan interface{}
+}
+
+func NewWriter(config *config.Config, sequencerRepoPath string) (self *Writer) {
+	self = new(Writer)
+	self.Output = make(chan interface{}, 1)
+
+	self.Task = task.NewTask(config, "writer").
+		WithSubtaskFunc(self.run)
```
Comment (janekolszak) [inline src/initsequencer/writer.go:40]: close the Output channel

File: src/initsequencer/writer.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	NUMBER_OF_NODES         = 3
+)
+
+type Writer struct {
+	*task.Task
+
+	sequencerRepoPath string
+	db                *gorm.DB
+	input             chan *arweave.Block
+	Output            chan interface{}
```
Comment (janekolszak) [inline src/initsequencer/writer.go:32]: chan struct{}

File: src/utils/task/task.go [hunk 8]
```diff
@@ -407,5 +407,13 @@ func (self *Task) StopWait() {
 	}
 
 	self.Log.Info("Stopped")
+}
 
+func (self *Task) WithStopChannel(channel chan interface{}) *Task {
+	self.WithSubtaskFunc(func() error {
+		<-channel
```
Comment (janekolszak) [inline src/utils/task/task.go:414]: Wait on Ctx.Done() as well. 
Call self.Stop() only if channel was triggered, not the context.

File: src/utils/task/task.go [hunk 9]
```diff
@@ -407,5 +407,13 @@ func (self *Task) StopWait() {
 	}
 
 	self.Log.Info("Stopped")
+}
 
+func (self *Task) WithStopChannel(channel chan interface{}) *Task {
```
Comment (janekolszak) [inline src/utils/task/task.go:412]: struct{}

File: src/initsequencer/controller.go (hunk 3 above)
Comment (szynwelski) [reply src/initsequencer/controller.go:47]: done

File: src/initsequencer/writer.go (hunk 6 above)
Comment (szynwelski) [reply src/initsequencer/writer.go]: changed

File: src/initsequencer/writer.go (hunk 7 above)
Comment (szynwelski) [reply src/initsequencer/writer.go]: changed

File: src/utils/task/task.go (hunk 8 above)
Comment (szynwelski) [reply src/utils/task/task.go]: done

File: src/utils/task/task.go (hunk 9 above)
Comment (szynwelski) [reply src/utils/task/task.go]: changed

**Review by janekolszak — APPROVED** [review]

**Pull request merged**